Qemu linux boot

daandemeyer opened the pull request:
This allows using qemu's direct linux boot to boot images directly without involving a bootloader. This boots faster compared to going through UEFI but more importantly, it allows booting images of architectures that don't support UEFI  in qemu.

I wonder if we'd want to add "linux" as an option for "--boot-protocols" instead of gating all the build time changes behind the --qemu-linux-boot option. If we'd do that, we could also forgo adding boot loader partitions and all the related stuff if only "linux" is enabled in "--boot-protocols".

lucasdemarchi commented:
@DaanDeMeyer  I like the idea of adding it to --boot-protocols. It would make much sense for testing kernel changes.

DaanDeMeyer commented:
Updated to add "linux" to --boot-protocols. We still allocate the ESP partition since kernel-install doesn't generate initrds otherwise. Renamed --qemu-linux-boot to --qemu-boot and it now takes one of the boot protocols from --boot-protocols. It's used to determine the boot protocol we instruct qemu to use. So you can build an image with all three boot protocols enabled and boot each individually with qemu.

bluca commented:
I like this, i have local hacks for the same purpose

behrmann commented:
I too have use for this, since I sometimes need to extract the initrd and kernel. The name "linux" for this is a bit generic (I didn't have an immediate idea what it meant), how about something like "external" or "extracted"?

Since I see some overlap, should this gone in after #923?

DaanDeMeyer commented:
This should go in after #923 indeed. I'm stealing the name from qemu itself, which calls this "Direct Linux Boot" (https://qemu-project.gitlab.io/qemu/system/linuxboot.html).  I don't think "external" or "extracted" would be all that obvious either.

behrmann commented:
Ok, if this is a mostly standard term, then nevermind.

DaanDeMeyer commented:
Rebased but this also depends on #1007 to work properly in all scenarios.

keszybz commented on mkosi/__init__.py:
```diff
@@ -4350,17 +4347,24 @@ def install_unified_kernel(
+                "--add-section", f".osrel={osrelease}",   "--change-section-vma", ".osrel=0x20000",
+                "--add-section", f".cmdline={cmdline}",   "--change-section-vma", ".cmdline=0x30000",
+                "--add-section", f".linux={root / kimg}", "--change-section-vma", ".linux=0x2000000",
+                "--add-section", f".initrd={initrd}",     "--change-section-vma", ".initrd=0x3000000",
```
FWIW, dracut varied this by arch: 0x3000000 by default, but 0x4000000 on arm64. Was there some discussion of this value?

DaanDeMeyer replied:
Hmm, I just copied the original values from the systemd-stub manpage. I figure we can keep using the same ones for now. If we ever run into an issue building aarch64 UEFI images we can update the defaults.

keszybz commented on mkosi/__init__.py:
```diff
@@ -7275,6 +7279,21 @@ def setup_netdev(args: MkosiArgs, root: Path, do_run_build_script: bool, cached:
+            run_workspace_command(args, root, ["dpkg-reconfigure", "dracut"])
+
+        for kver, kimg in gen_kernel_images(args, root):
+            run_workspace_command(args, root, ["kernel-install", "add", kver, Path("/") / kimg])
```
> `Path("/") / kimg`

That's ugly, and all the callers do this fixup. Please just make the path absolute in `gen_kernel_images()`.

DaanDeMeyer replied:
Frankly, I don't immediately see a cleaner way to do this since we need to support referring to the kernel image in the context of the image rootfs and in the context of the host rootfs. There's 3 options:

1. Return the path relative to the image rootfs. This is what we do now. It means we have to prefix with the "anchor" whenever trying to use the path, either `/` or `root`.
2. Return the absolute path in the image rootfs. This means we only have to prefix the path when using it in the context of the host rootfs. However, joining 2 absolute paths in python is not exactly intuitive

```
>>> from pathlib import Path
>>> a = Path("/usr")
>>> b = Path("/abc")
>>> a / b
PosixPath('/abc')
```

3. Return the absolute path in the host rootfs. This means we have to transform the path when trying to use it in the image rootfs context. This comes down to something like the following:

```
Path("/") / kimg.relative_to(root)
```

Both 2 and 3 don't seem super intuitive to me, so it seems better to have `gen_kernel_images()` return a path relative to the image rootfs and either prepend `/` or `root` as needed depending on the context.

keszybz replied:
OK.

keszybz commented on mkosi.md:
```diff
@@ -1129,6 +1133,13 @@ a machine ID.
+`QemuBoot=`, `--qemu-boot=`
+
+: When used with the `qemu` verb, this option sets the boot protocol to be used
+  by qemu. Can be set to either `uefi`, `bios` or `linux`. Note that a boot
```
Same?

keszybz reviewed (commented):
Looks reasonable. I doubt we'd get this entirely correct on the first try, so I'd just merge this and apply fixes later if we find something is missing.

behrmann commented on mkosi/__init__.py:
```diff
@@ -4458,6 +4460,70 @@ def extract_unified_kernel(
+            kimgabs = root / kimg
+            initrd = root / prefix / args.machine_id / kver / "initrd"
+
+        if kimgabs is None:
+            die("No kernel image found, can't extract.")
+        assert initrd is not None
```
```suggestion
        if kimgabs is None:
            die("No kernel image found, can't extract.")
        assert initrd is not None
```
That's a bit weird, no? For `kimgabs` we print a nice error and for `initrd` being `None` only an assertion error that would need to be figured out?

behrmann commented on mkosi/__init__.py:
```diff
@@ -4458,6 +4460,70 @@ def extract_unified_kernel(
+    do_run_build_script: bool,
+    for_cache: bool,
+    mount: Callable[[], ContextManager[None]],
+) -> Tuple[Optional[BinaryIO], Optional[BinaryIO]]:
```
Could this maybe be made more precise like this?
```suggestion
) -> Union[Tuple[BinaryIO], BinaryIO], Tuple[None, None]]:
```

This looks more ugly now, but in the future (3.10+) this will would actually be nicely readable
```python
tuple[BinaryIO, BinaryIO] | tuple[None, None]
```
(while also possibly being more precise)

behrmann reviewed (commented):
One serving of LGTM with a few nits on the side. :)
